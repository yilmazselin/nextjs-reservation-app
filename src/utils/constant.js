export const localeOptions = [
  {
    id: 0,
    key: "tr"
  },
  {
    id: 1,
    key: "en"
  }
]
