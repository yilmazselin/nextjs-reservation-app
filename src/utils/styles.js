export const padding = {
  xs: 8,
  sm: 16,
  md: 24,
  lg: 32,
  xl: 40,
};

export const fontSize = {
  sm: 12,
  nm: 14,
  md: 16,
  lg: 24,
  xl: 32,
};

export const colors = {
  black: "#E82223",
  dark: "#121212",
  gray: "#64738C",
  light: "#E6E6E6",
  red: "#E82223",
  green: "#0DAFC0",
  yellow: "#ECD03F",
  blue: "#0078E3",
  white: "#fff",
  disabled: "#BBC3CF",
  zero: "#000000",
};

export const fontWeight = {
  bold: 700,
  semiBold: 600,
  regular: 500,
  medium: 400,
  light: 300,
};
