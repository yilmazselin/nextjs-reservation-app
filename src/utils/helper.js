import { css } from "styled-components"
import { colors, fontSize, fontWeight } from "./styles"

const breakpoints = {
  sm: 576,
  md: 768,
  lg: 992,
  xl: 1200,
  xxl: 1400
}

export const containerWidths = {
  sm: "540px",
  md: "720px",
  lg: "960px",
  xl: "1140px",
  xxl: "1320px"
}

export const responsive = Object.keys(breakpoints).reduce(
  (accumulator, label) => {
    accumulator[label] = (...args) => css`
      @media (min-width: ${breakpoints[label]}px) {
        ${css(...args)};
      }
    `

    return accumulator
  },
  {}
)

export const font = ({ size, weight, color }) => `
  color: ${color
    ? color
    : colors.black};
  font-size: ${size
    ? size
    : fontSize.nm}px;
  font-family: "NunitoSans-${weight
    ? weight
    : fontWeight.medium}","sans-serif";
`
