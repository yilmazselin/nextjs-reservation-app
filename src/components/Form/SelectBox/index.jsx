import { memo } from "react"
import { useTranslation } from "next-i18next"
import styled from "styled-components"
import { font } from "../../../utils/helper"
import { colors, fontSize, fontWeight } from "../../../utils/styles"

const Container = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
  margin-bottom: ${(props) => props.marginBottom}px;
  overflow: hidden;
`

const Label = styled.label`
  ${font({
    size: fontSize.md,
    weight: fontWeight.regular,
    color: colors.gray
  })}
  margin-bottom:10px;
  font-size: 13px;
`

const Select = styled.select`
  border: none;
  outline: none;
  border-bottom: 1px solid ${colors.gray};
  padding-bottom: 8px;
  margin-left: -5px;
  width: calc(100% + 5px);
  ${font({
    size: fontSize.md,
    weight: fontWeight.semiBold,
    color: colors.dark
  })};
`

function SelectBox({
  id,
  label,
  name,
  marginBottom,
  options,
  onChange,
  defaultValue
}) {
  const { t } = useTranslation("common")

  return (
    <Container marginBottom={marginBottom}>
      <Label htmlFor={id}>{label}</Label>
      <Select
        data-testid="select"
        defaultValue={defaultValue}
        id={id}
        name={name}
        onChange={onChange}
      >
        {options
          ? options.map((x) => <option data-testid="select-option" key={x.id} value={x.key}>
            {t(`common:languages.${x.key}`)}
          </option>)
          : null}
      </Select>
    </Container>
  )
}

export default memo(SelectBox)
