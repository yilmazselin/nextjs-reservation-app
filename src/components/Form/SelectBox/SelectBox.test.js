import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import "jest-styled-components";
import { useTranslation } from "react-i18next";
import SelectBox from ".";
import { localeOptions } from "../../../utils/constant";

jest.mock("react-i18next", () => ({
  useTranslation: jest.fn(),
}));

const tSpy = jest.fn((str) => str);
const useTranslationSpy = useTranslation;

useTranslationSpy.mockReturnValue({
  t: tSpy,
  i18n: {
    changeLanguage: () => new Promise(() => {}),
  },
});

describe("<SelectBox />", () => {
  it("Should render correctly", () => {
    const { container } = render(<SelectBox />);

    expect(container).toMatchSnapshot();
  });

  it("Should have the correct label", () => {
    render(<SelectBox id="test-select" label="Test Select" />);
    expect(screen.getByLabelText("Test Select")).toBeInTheDocument();
  });

  it("Selection simulates", () => {
    render(<SelectBox options={localeOptions} />);

    expect(tSpy).toHaveBeenCalledTimes(2);

    fireEvent.change(screen.getByTestId("select"), { target: { value: "tr" } });

    const options = screen.getAllByTestId("select-option");

    expect(options[0].selected).toBeTruthy();
    expect(options[1].selected).toBeFalsy();
  });

  it("Should call changeHandler onChange on selectbox", () => {
    const mockHandler = jest.fn();

    render(<SelectBox onChange={mockHandler} options={localeOptions} />);
    fireEvent.change(screen.getByTestId("select"), { target: { value: "tr" } });
    expect(mockHandler.mock.calls.length).toBe(1);
  });
});
