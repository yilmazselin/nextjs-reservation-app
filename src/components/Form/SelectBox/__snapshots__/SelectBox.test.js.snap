// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectBox /> Should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  margin-bottom: px;
  overflow: hidden;
}

.c1 {
  color: #64738C;
  font-size: 16px;
  font-family: "NunitoSans-500","sans-serif";
  margin-bottom: 10px;
  font-size: 13px;
}

.c2 {
  border: none;
  outline: none;
  border-bottom: 1px solid #64738C;
  padding-bottom: 8px;
  margin-left: -5px;
  width: calc(100% + 5px);
  color: #121212;
  font-size: 16px;
  font-family: "NunitoSans-600","sans-serif";
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
    />
    <select
      class="c2"
      data-testid="select"
    />
  </div>
</div>
`;
